some additional region refactorings

lcnr opened the pull request:
the commits are selfcontained ✨ 

r? @BoxyUwU

lcnr committed:
replace unnecessary folder impls with fold_region

BoxyUwU commented:
r=me if CI passes

compiler-errors commented:
@bors r=BoxyUwU

bors commented:
:pushpin: Commit cf77639d27ba3dfb3c10c90762463e78e0476424 has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU cf77639d27ba3dfb3c10c90762463e78e0476424 -->
<!-- homu: {"type":"Approved","sha":"cf77639d27ba3dfb3c10c90762463e78e0476424","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #117979) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr committed:
rename bound region instantiation

- `erase_late_bound_regions` -> `instantiate_bound_regions_with_erased`
- `replace_late_bound_regions_X` -> `instantiate_bound_regions_X`

lcnr commented:
added cargo to the second commit by accident

bors commented:
:pushpin: Commit 40b154e53c0e04ff4cfd40d43d8e2b86b143b763 has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU 40b154e53c0e04ff4cfd40d43d8e2b86b143b763 -->
<!-- homu: {"type":"Approved","sha":"40b154e53c0e04ff4cfd40d43d8e2b86b143b763","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 40b154e53c0e04ff4cfd40d43d8e2b86b143b763 with merge 069a4af48b4b2111ff65f3895edb425683e6ae8a...
<!-- homu: {"type":"BuildStarted","head_sha":"40b154e53c0e04ff4cfd40d43d8e2b86b143b763","merge_sha":"069a4af48b4b2111ff65f3895edb425683e6ae8a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6905247679/job/18791719750)
Approved by: BoxyUwU
Pushing 069a4af48b4b2111ff65f3895edb425683e6ae8a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"BoxyUwU","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6905247679/job/18791719750"},"merge_sha":"069a4af48b4b2111ff65f3895edb425683e6ae8a"} -->

rust-timer commented:
Finished benchmarking commit (069a4af48b4b2111ff65f3895edb425683e6ae8a): [comparison URL](https://perf.rust-lang.org/compare.html?start=4d7f952a02d0bca67c98a6b74895b7e3fbe38341&end=069a4af48b4b2111ff65f3895edb425683e6ae8a&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=4d7f952a02d0bca67c98a6b74895b7e3fbe38341&end=069a4af48b4b2111ff65f3895edb425683e6ae8a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.1%, 0.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.3%, 1.0%] | 10    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.1%, 0.1%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4d7f952a02d0bca67c98a6b74895b7e3fbe38341&end=069a4af48b4b2111ff65f3895edb425683e6ae8a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.5%, 1.1%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.3%, 3.6%]   | 4     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-1.3%, -0.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 2     |
| All ❌✅ (primary)                 | -0.2% | [-1.3%, 1.1%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4d7f952a02d0bca67c98a6b74895b7e3fbe38341&end=069a4af48b4b2111ff65f3895edb425683e6ae8a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.5% | [2.3%, 2.8%] | 3     |
| Regressions ❌ <br /> (secondary)  | 4.1% | [2.0%, 6.8%] | 8     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.5% | [2.3%, 2.8%] | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4d7f952a02d0bca67c98a6b74895b7e3fbe38341&end=069a4af48b4b2111ff65f3895edb425683e6ae8a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.352s -> 676.162s (-0.03%)
**Artifact size**: 313.63 MiB -> 313.62 MiB (-0.00%)

Mark-Simulacrum commented:
Regressions appear to be genuine but are mostly limited to secondary benchmarks. Not worth further investigation, I think.
